### ISSUE TITLE
Custom importer sass.NULL usage

### ISSUE THREAD

**Issue by kenborge** [sass/node-sass#1291]
The documentions for custom importers state that 
```If an importer does not want to handle a particular path, it should return sass.NULL```

But if you implement the importer in a package, that will be used as a dependency, this package will have a dependency on node-sass and have node-sass package in it's node_modules.

So when this importer is run from another package, like grunt-sass that has its own dependency on node-sass, the node-sass module used in grunt-sass and in the importer will not be the same object and the object reference equality check in node-sass will fail.

```javascript 
var result = importer.call(options.context, file, prev, done);
//result will be a sass.NULL from another node-sass object and the === fails
if (result) {
    done(result === module.exports.NULL ? null : result);
}
```

This cause node-sass to think all imports are handled by this importer no matter what.

**Comment by kenborge** [comment]
This is assuming the importer implementation does a require('node-sass') to get a node-sass reference. How else would it be able to access sass.NULL

**Comment by xzyfer** [comment]
Good questions. Maybe @am11 or @saper would have thoughts here.

**Comment by kenborge** [comment]
In some situation using peerDependencies will probably solved it, but probably not everywhere

**Comment by saper** [comment]
Would that help if we would just return undefined? I actually think we should drop sass.NULL and use that native node type, just haven't got around to implementing it yet.

**Comment by xzyfer** [comment]
@chriseppstein I'd like to hear your thoughts on this since I expect any decision here will impact eyeglass.

**Comment by am11** [comment]
@kenborge, you have made a very good point and thanks for spotting and reporting this issue.
This issue is in node-sass code itself, which can be hot fixed as:

```diff
- done(result === module.exports.NULL ? null : result);
+ done(result.toString() === module.exports.NULL.toString() ? null : result);
// both will return `[object SassNull]` which is different than `[object Object]`
```

or

```diff
- done(result === module.exports.NULL ? null : result);
+ done(result.constructor.toString() === module.exports.NULL.constructor.toString() ? null : result);
```

The idea was to not advertise these export too much and still able to provide control for explicitly returning `SassNull` type.

**Comment by saper** [comment]
@kenborge can you check if my preliminary fix #1308 makes it better? Are you importers called synchronously or asynchronously? If async, then this could be a duplicate of #1296.

**Comment by saper** [comment]
Actually asking for `sass.NULL` is a mistake. With #1308 returning JavaScript `null` should work as well.

**Comment by am11** [comment]
In JavaScript `return null;`, `return;` and returning nothing are interchangeable, which is not applicable here as we need to differentiate between `done` and `return`. Explicitly returning `sass.NULL` is by design.

**Comment by saper** [comment]
@am11 What about `undefined`?

**Comment by am11** [comment]
@saper, either test the type `sass.NULL` or drop `sass.NULL` completely and use `return null` everywhere (i.e. breaking change). The reason why it wasn't done that (`return null` and test based on `undefined`) way in first place is to avoid the confusion and keep the type-systems separate.

**Comment by saper** [comment]
My patch uses `null` everywhere and aliases `sass.NULL` and `sass.sass_types.Null.NULL` to `null` for compatibility.

**Comment by xzyfer** [comment]
A couple things.

> if you implement the importer in a package, that will be used as a dependency, this package will have a dependency on node-sass

This is essentially forcing a peer dependency, which I agree is bad.

>Actually asking for sass.NULL is a mistake.
>Explicitly returning sass.NULL is by design.

I agree that with returning `sass.NULL`.

> In JavaScript return null;, return; and returning nothing are interchangeable

Technically not true.

```
> function foo() { return null; }
undefined
> function bar() { return; }
undefined
> foo()
null
> bar()
undefined
```

> What about undefined?

We could strictly check against `undefined` to get around the issue @am11 mentions with return `null`. @saper's patch does this.

Although technically `null` is not `undefined` as @am11 points out they are used interchangeably in the JS community. So much so a lot of developers don't know the conceptual difference. This means we're practically asking for false bug reports due to this misunderstanding.

****

My preference would be move sass-types into it's own repo as discussed in https://github.com/sass/node-sass/issues/1088#issuecomment-166467651

This solves the peerDependency problem, preserves the semantics of returning `sass.NULL` from importers/functions, and resolves abilities with returning null vs undefined.

**Comment by xzyfer** [comment]
It's also worth noting that we currently do not distinguish between custom importers returning `false`, `null`, `undefined`, or `sass.NULL`.

The exception being #1308 where we don't correctly handle return `sass.NULL` with async importers.

**Comment by xzyfer** [comment]
I've added test coverage for the existing importer noop return semantics in #1319. This will prevent us from unintentionally introducing BC breaks.

**Comment by xzyfer** [comment]
TLDR;

@kenborge you issue is fixed in https://github.com/sass/node-sass/pull/1318 and will be in v3.4.3.
@am11 @saper I've opened #1320 to discussion creating a `sass-types` package.

**Comment by saper** [comment]
There are several problems with explicitly using `undefined` (it can be redefined for example), but we use it only because we want to be nice to those returning via `return value` instead of `done(value)` - we intercept that return value for them. It is not a proper solution, but kind of works and checking for `undefined` fits that ugly pattern.

**Comment by saper** [comment]
Maybe we would better off by moving this code into the C++ binding. We should know better in C++ if the value was returned and how.

**Comment by pmowrer** [comment]
@xzyfer Returning `null` or `undefined` in place of `sass.NULL` does not appear to work (`3.4.2`). The importer will just hang indefinitely after processing a few files whereas it completes using `sass.NULL`.

**Comment by xzyfer** [comment]
@pmowrer I believe this has been patched in the current beta. Try `npm install node-sass@beta`.
